src/parser/jobs/sch/DoTs.js

```diff
@@ -0,0 +1,134 @@
+	static dependencies = [
+		'checklist',
+		'combatants',
+		'cooldowns',
+		'enemies',
+		'invuln',
+		'suggestions',
+	]
+
+	_lastApplication = {}
+	_clip = {
+		[STATUSES.BIO_II.id]: 0,
+		[STATUSES.MIASMA.id]: 0,
+	}
+
+	constructor(...args) {
+		super(...args)
+
+		const filter = {
+			by: 'player',
+			abilityId: [STATUSES.BIO_II.id, STATUSES.MIASMA.id],
+		}
+		this.addHook(['applydebuff', 'refreshdebuff'], filter, this._onDotApply)
+		this.addHook('complete', this._onComplete)
+	}
+
+	_onDotApply(event) {
+		const statusId = event.ability.guid
+
+		// Make sure we're tracking for this target
+		const lastApplication = this._lastApplication[event.targetID] = this._lastApplication[event.targetID] || {}
+
+		// If it's not been applied yet, set it and skip out
+		if (!lastApplication[statusId]) {
+			lastApplication[statusId] = event.timestamp
+			return
+		}
+
+		// Base clip calc
+		let clip = STATUS_DURATION[statusId] - (event.timestamp - lastApplication[statusId])
+
+		// Remove any untargetable time from the clip - often want to hardcast after an invuln phase, but refresh w/ 3D shortly after.
+		clip -= this.invuln.getUntargetableUptime('all', event.timestamp - STATUS_DURATION[statusId], event.timestamp)
+
+		// Also remove invuln time in the future that casting later would just push dots into
+		// TODO: This relies on a full set of invuln data ahead of time. Can this be trusted?
+		clip -= this.invuln.getInvulnerableUptime('all', event.timestamp, event.timestamp + STATUS_DURATION[statusId] + clip)
+
+		// Capping clip at 0 - less than that is downtime, which is handled by the checklist requirement
+		this._clip[statusId] += Math.max(0, clip)
+
+		lastApplication[statusId] = event.timestamp
+	}
+
+	_onComplete() {
+		// Checklist rule for dot uptime
+		this.checklist.add(new Rule({
+			name: 'Keep your DoTs up',
+			description: <Fragment>
+				As a Scholar, DoTs are a significant portion of your sustained damage. Aim to keep them up at all times.
+			</Fragment>,
+			requirements: [
+				new Requirement({
+					name: <Fragment><ActionLink {...ACTIONS.BIO_II} /> uptime</Fragment>,
+					percent: () => this.getDotUptimePercent(STATUSES.BIO_II.id),
+				}),
+				new Requirement({
+					name: <Fragment><ActionLink {...ACTIONS.MIASMA} /> uptime</Fragment>,
+					percent: () => this.getDotUptimePercent(STATUSES.MIASMA.id),
+				}),
+				new Requirement({
+					name: <Fragment><ActionLink {...ACTIONS.SHADOW_FLARE}/> uptime</Fragment>,
+					percent: () => this.getShadowFlareUptimePercent(),
+				}),
+			],
+		}))
+
+		// Suggestion for DoT clipping
+		const maxClip = Math.max(...Object.values(this._clip))
+		this.suggestions.add(new Suggestion({
+			icon: ACTIONS.BIO.icon,
+			content: <Fragment>
+				Avoid refreshing DoTs significantly before their expiration, except when at the end of the fight. Unnecessary refreshes risk overwriting buff snapshots, and increase the frequency you&apos;ll need to hardcast your DoTs.
```

Comment:
This suggestion is primarily worded for SMN, where you want to keep DoTs up with Tri-Disaster as much as possible. Might need some tweaks.